pkg/discovery/dtofactory/workload_controller_entity_dto_builder.go

```diff
@@ -0,0 +1,180 @@
+package dtofactory
+
+import (
+	"github.com/golang/glog"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/repository"
+	"github.com/turbonomic/kubeturbo/pkg/util"
+	sdkbuilder "github.com/turbonomic/turbo-go-sdk/pkg/builder"
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+)
+
+type workloadControllerDTOBuilder struct {
+	kubeControllersMap map[string]*repository.KubeController
+	namespaceUIDMap    map[string]string
+}
+
+func NewWorkloadControllerDTOBuilder(kubeControllersMap map[string]*repository.KubeController,
+	namespaceUIDMap map[string]string) *workloadControllerDTOBuilder {
+	return &workloadControllerDTOBuilder{
+		kubeControllersMap: kubeControllersMap,
+		namespaceUIDMap:    namespaceUIDMap,
+	}
+}
+
+// Build entityDTOs based on the given map from controller UID to KubeController entity.
+func (builder *workloadControllerDTOBuilder) BuildDTOs() ([]*proto.EntityDTO, error) {
+	var result []*proto.EntityDTO
+	for _, kubeController := range builder.kubeControllersMap {
+		// Id
+		workloadControllerId := kubeController.UID
+		entityDTOBuilder := sdkbuilder.NewEntityDTOBuilder(proto.EntityDTO_WORKLOAD_CONTROLLER, workloadControllerId)
+
+		// Display name
+		workloadControllerDisplayName := kubeController.Name
+		entityDTOBuilder.DisplayName(workloadControllerDisplayName)
+
+		// Resource commodities sold
+		commoditiesSold, err := builder.getCommoditiesSold(kubeController)
+		if err != nil {
+			glog.Errorf("Error creating commoditiesSold for %s: %v", kubeController.GetFullName(), err)
+			continue
+		}
+		entityDTOBuilder.SellsCommodities(commoditiesSold)
+
+		// Resource commodities bought from namespace
+		namespaceUID, exists := builder.namespaceUIDMap[kubeController.Namespace]
+		if exists {
+			commoditiesBought, err := builder.getCommoditiesBought(kubeController, namespaceUID)
+			if err != nil {
+				glog.Errorf("Error creating commoditiesBought for %s: %v", kubeController.GetFullName(), err)
+				continue
+			}
+			entityDTOBuilder.Provider(sdkbuilder.CreateProvider(proto.EntityDTO_NAMESPACE, namespaceUID)).BuysCommodities(commoditiesBought)
+			// Set movable false to avoid moving WorkloadController across namespaces
+			entityDTOBuilder.IsMovable(proto.EntityDTO_NAMESPACE, false)
+		} else {
+			glog.Errorf("Failed to get namespaceUID from namespace %s for controller %s", kubeController.Namespace,
+				kubeController.GetFullName())
+		}
+
+		// Create WorkloadControllerData to store controller type data
+		entityDTOBuilder.WorkloadControllerData(builder.createWorkloadControllerData(kubeController))
```

Comment:
I am trying to find the answer by reading the review but I still couldn't figure out. My question here is , how are we dealing with pods under replicaset then under deployment. Are we recursively find the owner and only create one workloadcontroller in this case we we discussed? I couldn't find the related logic in this review.

Reply:
Sorry I didn't explain this clearly in the description. Yes we create one WorkloadController with controller type as "Deployment" in this case.

We have already had existing logic to get the pod owner in `getPodOwner` func in `cluster_monitor.go` by calling `GetPodGrandInfo` func, which returns "Deployment" as the owner "type" of the pod in this case instead of "ReplicaSet". Then we store the pod owner metrics (owner name, type and UID) in the metrics sink by calling `genOwnerMetrics` in `cluster_monitor.go`.

In the changes here in this PR, `KubeController` entity is created by retrieving pod owner name, type and UID from metrics sink by calling `getKubeControllerInfo` func in `controller_metrics_collector.go`. The owner type (controller type) retrieved here is already "Deployment" in this case. Eventually, we convert `KubeController` entity to WorkloadController entity DTO in `workload_controller_entity_dto_builder.go`.

Reply:
Thanks Yue, that makes a lot of sense.